src/state/selection/logics.ts

```diff
@@ -1,46 +1,73 @@
 import { createLogic } from "redux-logic";
-import { UIDisplayData } from "@aics/simularium-viewer";
 
 import { ReduxLogicDeps } from "../types";
-import { getSimulariumFile } from "../trajectory/selectors";
+import {
+    getDefaultUIDisplayData,
+    getSimulariumFile,
+} from "../trajectory/selectors";
+import { getCurrentUIData } from "../compoundSelectors";
+import {
+    GET_DISPLAY_DATA_FROM_BROWSER,
+    HANDLE_COLOR_CHANGE,
+} from "./constants";
+import { setCurrentColorSetting, setSelectedUIDisplayData } from "./actions";
+import { applyColorChangeToUiDisplayData, isSameAgentTree } from "../../util";
+import { ColorSetting } from "./types";
 
-import { APPLY_USER_COLOR } from "./constants";
-import { getSelectedUIDisplayData } from "./selectors";
-import { setSelectedUIDisplayData } from "./actions";
-
-const storeColorsLogic = createLogic({
-    process(deps: ReduxLogicDeps, dispatch, done) {
-        const { action, getState } = deps;
-        const uiData: UIDisplayData = getSelectedUIDisplayData(getState());
+const handleColorChangeLogic = createLogic({
+    validate(deps: ReduxLogicDeps, allow, reject) {
+        const { getState, action } = deps;
         const colorChange = action.payload;
-        const newUiData = uiData.map((agent) => {
-            const newAgent = { ...agent };
-            if (agent.name === colorChange.agent.name) {
-                if (colorChange.agent.tags.includes("")) {
-                    newAgent.color = colorChange.color;
-                }
-                const newDisplayStates = agent.displayStates.map(
-                    (state: any) => {
-                        if (colorChange.agent.tags.includes(state.id)) {
-                            return {
-                                ...state,
-                                color: colorChange.color,
-                            };
-                        }
-                        return state;
-                    }
-                );
-                newAgent.displayStates = newDisplayStates;
-            }
-            return newAgent;
+        const currentDisplayData = getCurrentUIData(getState());
+        const newDisplayData = applyColorChangeToUiDisplayData(
+            colorChange,
+            currentDisplayData
+        );
+        if (newDisplayData !== currentDisplayData) {
+            const fileKey = getSimulariumFile(getState()).name;
+            localStorage.setItem(fileKey, JSON.stringify(newDisplayData));
+            allow(setSelectedUIDisplayData(newDisplayData));
+        } else {
+            reject(action);
+        }
+    },
+    process() {
+        return setCurrentColorSetting({
+            currentColorSetting: ColorSetting.UserSelected,
```

Comment:
I haven't been making use of these execution phase hooks (`transform`/`validate`) or necessarily using the `process` hook very intentionally. I think this is a nice improvement and they definitely help clarify what's going on in cases where we may swap one action for another or dispatch actions if another one is successful.